WIP: Tracking Branch for changes to be upstreamed

fejfighter opened the pull request:
Do Not Merge,
This is the branch I have been working on, I will split it out in to smaller requests.

I just wanted to track what needs to be pushed up and to demonstrate what I have been working on

This does get child frame positioning mostly working, works for X, wayland and broadway, but not Xwayland.

I have managed to get this working under gtk-broadway (aside from a few key bindings that Firefox handled instead) which shows that we are getting close to pure implementation.

the one thing I have noticed that we are missing is handling clicks on the mode bar and buffer, which I have not had a chance to look at.

Commit:
add pgtk protection

Commit:
Some work toward posframe on wayland

hacky GTK offsets taht will need better calculations

Commit:
Moving the frame at early creation time has no effect

Commit:
End Resize flickering by copying surface rather than just clearing

Commit:
Don't queue redraw on update finish, it's not required

Commit:
gdk_flush is deprecated

Commit:
Bring pgtk more inline with X11-cairo builds

Commit:
Don't create a context each time, just hold a reference

Commit:
Pull in a few cairo image changes

Commit:
Remove redundant code in pgtk_update_begin()

Commit:
Remove pgtkconn, we really didn't need it

fejfighter committed:
Remove re-defined keymaps used for NS/macos ports

Commit:
Get parent frame's editor widget allocation for the offset

fejfighter committed:
Logging cleanup

Commit:
Address compiler warnings

Commit:
Set screen/display dpi res to make broadway work as expected

Commit:
Handle Fullscreen correctly, hiding title bars

fejfighter committed:
Fix some HiDPI stuff

Commit:
Fix child-frame offsets for negative values

Commit:
Fix compile Issue

masm11 commented:
@fejfighter 
> the one thing I have noticed that we are missing is handling clicks on the mode bar and buffer, which I have not had a chance to look at.

What problem is it?

When I click in a buffer, the cursor jumps to that point.
When I click on a buffer name on a mode line, buffer switches.

I see no problem.

fejfighter commented:
Sorry I was writing that in a bit of a rush:

right clicking in the buffer, and clicking on mini-modes don't bring up the context menu for the minibuffer.

admittedly, I never use these features, and we could potentially push this up without it implement then seek help from a wider developer community

the rest does work perfectly

masm11 commented:
right click! I forgot it...

fejfighter committed:
remove debug string

fejfighter commented:
!14 !15  and !16 are up and seem to be isolated changes.

16 is a small cleanup fix, the other 2 are a little more involved.

Sorry for dumping everything at once, I had meant to split this out closer to writing the code but I was a little lazy and didn't get back to it.

Thanks

fejfighter committed:
Fix some compiler warnings in pgtkmenu.c

there is still one memcopy warning that also appears in the XGTK version of
emacs, so leave that for later

fejfighter committed:
Fix up duplicate cairo context creation

akirakyle commented:
Sorry to hijack this PR but it seems there's no issue tab on this repo (perhaps because it's a fork?). I was wondering if there are any plans to try to upstream this work into gnu emacs proper? I originally thought this PR was for tracking the patches that were to go to the official gnu emacs upstream but now I'm pretty sure that @fejfighter is taking upstream to mean this fork.

I think there are many users out there who are trying to run wayland only systems but are forced to keep xwayland around for the ever decreasing programs still on x11 such as emacs. Then there are users like myself who are still completely on x11 running an x11 window manager such as [exwm](https://github.com/ch11ng/exwm) because there isn't [anything comparable on wayland yet](https://github.com/ch11ng/exwm/issues/559). However I'm starting to see improvements to other programs such as [firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1610199) target mainly wayland and I think the situation will only get worse with time. So I think having a pure gtk emacs is very important for the future of emacs.

It seems this project was mentioned on the [emacs-devel list a bit ago](https://lists.gnu.org/archive/html/emacs-devel/2019-04/msg01014.html) but got no response there so it seems there aren't many core emacs devs who are aware of the work going on here. The idea of a pure gtk emacs does get [brought up occasionally ](https://lists.gnu.org/archive/html/emacs-devel/2018-03/msg00157.html) and most recently it seems [there's someone else also working on this](https://lists.gnu.org/archive/html/emacs-devel/2020-04/msg01357.html) but I couldn't find any code. While I understand there's a high (perhaps unreasonably so) barrier to getting code into the official gnu emacs repo, it certainly would make the work that is being done here live on and benefit everyone else using emacs and struggling with this wayland transition.

masm11 commented:
@akirakyle 

> I was wondering if there are any plans to try to upstream this work into gnu emacs proper?

Yes, I want to merge this fork to mainline, but I have one more work to do before that.

However, it is not essential for English users, so I'll publish to emacs-devel soon.

fejfighter commented:
I got halfway through a reply before getting distracted.

I'd also be interested to see this go up.

@akirakyle, You're right, this PR was for tracking upstream to masm11, we might need to start a new one for gnu upstream.

masm11 commented:
@fejfighter @akirakyle 

https://lists.gnu.org/archive/html/emacs-devel/2020-04/msg01641.html

:)

akirakyle commented:
Fantastic! Thanks for starting this process!

fejfighter commented:
@masm11 ill try to look at the mem issue mentioned on emacs-devel some time tonight.

fejfighter commented:
#22 up. should handle the compilation error mentioned on emacs-devel

fejfighter commented:
I think all the changes are in. no need to keep this open
